File: engine/src/main/java/org/epdb/engine/comparison/ComparisonPredicate.java
package org.epdb.engine.comparison;

import org.epdb.engine.dto.ColumnValue;
import org.epdb.engine.dto.Tuple;

public record ComparisonPredicate(
        int columnIndex,
        Op operator,
        ColumnValue constantValue
) implements Predicate {

    @Override
    public boolean evaluate(Tuple tuple) {
        var columnValue = tuple.getValueAtIndex(this.columnIndex);
        if (!(this.constantValue instanceof ColumnValue constantColumnValue)) {
            return false;
        }

        try {
            int comparison = columnValue.comparesTo(constantColumnValue);
            return this.operator.apply(comparison);
        } catch (Exception e) {
            return false;
        }
    }
}

File: engine/src/main/kotlin/org/epdb/engine/comparison/ComparisonPredicate.kt
package org.epdb.engine.comparison

import org.epdb.engine.columntypes.ColumnValue
import org.epdb.engine.dto.Tuple
import org.epdb.engine.comparisonoperator.Operator

data class ComparisonPredicate(
    val columnIndex: Int,
    val operator: Operator,
    val constantValue: ColumnValue
) : Predicate {

    override fun evaluate(tuple: Tuple): Boolean {
        val columnValue = tuple.getValueAtIndex(columnIndex)?: return false
        val comparison = columnValue.comparesTo(constantValue)

        return operator.apply(comparison)
    }
}
